### PR TITLE
[RFC] ping watchdog during slow operations

### PR DESCRIPTION
**Pull request by vcaputo**
The core manager event loop can be blocked for potentially long durations, especially for vfs-heavy operations on busy/slow storage.

With this change we can start updating the watchdog on such slow-but-progressing operations, until an upper bound to still let the watchdog timeout expire on infinite loop scenarios.

### CODE REVIEW

**Comment by vcaputo** [comment]
The way I look at it is the service manager process has too much blocking functionality piled into it.  When it runs off on one of these hugely variable tasks dependent on external, highly variable factors like IO performance+utilization, like listing all the units can be, the watchdog timeout should become elastic temporarily *there*.  Because the PID1 has become something other than strictly a service manager at that moment, which is really the root of the problem IMO.

For the vast majority of PID1's execution it's just being init; reaping zombies and starting/stopping things.  The watchdog as-implemented works fine for those conditions, and is both easy *and* desirable to configure the timeout to be rather tight during, to reboot quickly on real hangs.

But such an arrangement blows up, often unexpectedly, when PID1 runs off into la-la land doing stuff that arguably shouldn't really be in-process.  The compromise of adding some explicit watchdog pings on such code paths where we *know* we're running off and doing arguably inappropriate amounts of work in-process seems reasonable to me.  But it requires at least a tacit acknowledgement that "hey guys, these operations really shouldn't happen in the service manager event loop" which is what those manager_ping_watchdog() calls are really stating in writing.

TL;DR when PID1 grew long tentacles of blocking execution complexity, it created the need for this.

**Comment by vcaputo** [comment]
> And I agree that the most useful answer the watchdog can give is, "is pid1 still responding to external calls" - ie, can it shut down the system, can it react to crashes/restarts, etc. If we are stuck reading from a filesystem, even if it's doing some nominal progress, it still doesn't answer a useful question.

Except you're defining the entire host's ability to do useful work by PID1's responsiveness, when those are very different things, no?

If I have a host running memory-resident services that are functioning perfectly well, having many active users connected to said services, but I happened to be doing administration work generating a bunch of IO the system never otherwise experiences, and make the mistake of running `systemctl list-unit-files` and poof the hw watchdog reboots the system - is that right?

Maybe the argument is really more along the lines of "why doesn't `systemctl list-unit-files` get the list of units from PID1, and do all the VFS-heavy work by itself, merging the results for enabled vs. disabled statuses?" and I think the obvious answer is: PID1 implements the dbus API.

Either way, something smells here.

**Comment by bluca** [comment]
> > And I agree that the most useful answer the watchdog can give is, "is pid1 still responding to external calls" - ie, can it shut down the system, can it react to crashes/restarts, etc. If we are stuck reading from a filesystem, even if it's doing some nominal progress, it still doesn't answer a useful question.
> 
> Except you're defining the entire host's ability to do useful work by PID1's responsiveness, when those are very different things, no?

Not really - PID1 is the system's manager, so it is important to be able to answer the question of whether it's ready to do such management work, or if it is stuck.
Let me flip it on its head: is it really useful for the watchdog to never fire if, somehow, we get stuck in that loop? Is that what the expectations around this watchdog are? We can't start/stop/restart services, cannot answer queries about the system state, cannot start event-triggered services, none of that works anymore, but some internal loop that reads is still running. Is the system really in a 'healthy' state?

> If I have a host running memory-resident services that are functioning perfectly well, having many active users connected to said services, but I happened to be doing administration work generating a bunch of IO the system never otherwise experiences, and make the mistake of running `systemctl list-unit-files` and poof the hw watchdog reboots the system - is that right?
> 
> Maybe the argument is really more along the lines of "why doesn't `systemctl list-unit-files` get the list of units from PID1, and do all the VFS-heavy work by itself, merging the results for enabled vs. disabled statuses?" and I think the obvious answer is: PID1 implements the dbus API.

Because the interesting answer is what the system manager's current understanding of the units is, rather than what is on disk. The two are not necessarily the same.

> Either way, something smells here.

I've been where you are literally two weeks ago. The answer is: the configuration smelled, and it was set to the 'best case scenario' rather than the worst case (but still tolerable) scenario. It did not take into account very real, and recoverable, temporary slowness. I looked at exactly the same things that you looked at, but in the end the answer was simply to change the configuration to adapt it to the specific use case.

**Comment by poettering** [comment]
> Maybe the argument is really more along the lines of "why doesn't `systemctl list-unit-files` get the list of units from PID1, and do all the VFS-heavy work by itself, merging the results for enabled vs. disabled statuses?" and I think the obvious answer is: PID1 implements the dbus API.

It can actually do this client-side. Just run the tools with $SYSTEMCTL_INSTALL_CLIENT_SIDE=1 env var set.

But this possibly creates these synchronization issues I mentioned, and of course requires full privs client side (i.e. no polkit for you, sorry) and direct access to the fs, selinux blablabla

**Comment by bluca** [comment]
> > Edit: also, to Lennart's point about why involve userspace with the watchdog ping at all - just let the kernel do it; It's a classic UNIX style "put the policy in userspace, we just implement the mechanism" scenario. I don't personally think it was ever intended that the hw watchdog be wired directly to userspace's ability to list a configuration-and-load-dependent amount of files on storage via dbus. At it's core it's still primarily a mechanism for rebooting crashed boxen.
> 
> But what is a "crashed box"? Quite often on Linux it's not the full system that freezes anymore, but just parts of it, i.e. some interrupt gets stuck, or some kernel assertion is ihit by _one_ thread of the system. These issues might propagate elsewhre and cause more stuff to get stuck, but "crashed box" is not a binary thing these days. Hence, from my PoV we should make sure that every component of the userspace OS is supervised by _something_. And the way we answered that for systemd is to say: PID 1 supervises services, and the hw watchdog supervises PID 1. This way if something gets stuck then it's covered, and can be dealt with in a somewhat local way, if possible at all. There are certainly other ways to design this though, but at some point it's up to us to decide what we bind the watchdog ping to, and yeah, this is how we did it.

And in the end, if someone does _not_ want these guarantees for their system for whatever reason, but they just want a dumb configuration and ping-in-a-loop mechanism, they don't have to use systemd for this, they can disable the configuration here and switch to "ye olde watchdogd" which is exactly that and nothing more.
